lib/services/server.go

```diff
@@ -370,6 +374,42 @@ func (s *ServerV2) CheckAndSetDefaults() error {
 	return nil
 }
 
+// DeepCopy creates a clone of this server value
+func (s *ServerV2) DeepCopy() Server {
+	return proto.Clone(s).(*ServerV2)
+}
+
+// Merge overwrites r from src and
+// is part of support for cloning Server values
+// using proto.Clone.
+// Implements proto.Merger
+func (r *Rotation) Merge(src proto.Message) {
+	s, ok := src.(*Rotation)
+	if !ok {
+		return
+	}
+	*r = *s
```

Comment:
The `proto.Merger` docs say:
```
	// Merge merges the contents of src into the receiver message.
	// It clones all data structures in src such that it aliases no mutable
	// memory referenced by src.
```
There are 2 possible issues:
- if `r.Foo` is set and `s.Foo` is empty, I assume that `r.Foo` value should be preserved
- if `Rotation` ever gets a pointer field, we break the promise of not keeping mutable memory references

I don't fully understand why we need to implement `proto.Merger`. Does `proto.Clone` not work on `Rotation` otherwise?

Reply:
Can you link to the docs you reference, btw?
I know this is not the complete Merger implementation - it implements enough to support the Clone use-case so it assumes that the receiver is zero value. Unfortunately I was not able to find a concise way to implement it to fulfill the Merger requirements you described and yet be able to avoid exhaustive manual tests of each attribute which is exactly what I was trying to achieve with `Clone`.
Also, the [Merger](https://github.com/gogo/protobuf/blob/v1.3.1/proto/clone.go#L58-L60) interface used here has a somewhat less strict requirement:
```
// Merge merges src into this message.
// Required and optional fields that are set in src will be set to that value in dst.
// Elements of repeated fields will be appended.
// ...
```

As to why I implemented `Merger` also for `Rotation` - exactly for the same reason as for `Metadata` - b/c it was not working with reflect due to use of `time.Time` which sports unexported fields.

Reply:
Ahh, I was looking at the regular protobuf docs, not gogoproto: https://github.com/golang/protobuf/blob/master/proto/proto.go#L99-L101

I think this merger implementation will cause problems later on, when it starts getting hit from non-Clone use-cases for whatever reason.
For now, it's OK merge to fix the race, but leave a comment explaining that it's a partial implementation. Next person reading this code will appreciate the background :)